core/src/main/java/com/linecorp/armeria/internal/common/KeepAliveHandler.java

```diff
@@ -0,0 +1,399 @@
+        }
+        if (pingIntervalMillis <= 0) {
+            pingIdleTimeNanos = 0;
+        } else {
+            pingIdleTimeNanos = TimeUnit.MILLISECONDS.toNanos(pingIntervalMillis);
+        }
+    }
+
+    public final void initialize(ChannelHandlerContext ctx) {
+        // Avoid the case where destroy() is called before scheduling timeouts.
+        // See: https://github.com/netty/netty/issues/143
+        if (isInitialized) {
+            return;
+        }
+        isInitialized = true;
+
+        if (connectionIdleTimeNanos > 0) {
+            connectionIdleTimeout = executor().schedule(new ConnectionIdleTimeoutTask(ctx),
+                                                        connectionIdleTimeNanos, TimeUnit.NANOSECONDS);
+        }
+        if (pingIdleTimeNanos > 0) {
+            pingIdleTimeout = executor().schedule(new PingIdleTimeoutTask(ctx),
+                                                  pingIdleTimeNanos, TimeUnit.NANOSECONDS);
+        }
+    }
+
+    public final void destroy() {
+        isInitialized = true;
+        if (connectionIdleTimeout != null) {
+            connectionIdleTimeout.cancel(false);
+            connectionIdleTimeout = null;
+        }
+        if (pingIdleTimeout != null) {
+            pingIdleTimeout.cancel(false);
+            pingIdleTimeout = null;
+        }
+        pingState = PingState.SHUTDOWN;
+        cancelFutures();
+    }
+
+    public final void onReadOrWrite() {
+        if (pingState == PingState.SHUTDOWN) {
+            return;
+        }
+
+        if (connectionIdleTimeNanos > 0) {
+            lastConnectionIdleTime = System.nanoTime();
+            firstConnectionIdleEvent = true;
+        }
+
+        if (pingResetsPreviousPing()) {
+            if (pingIdleTimeNanos > 0) {
+                lastPingIdleTime = System.nanoTime();
+                firstPingIdleEvent = true;
+            }
+            pingState = PingState.IDLE;
+            cancelFutures();
+        }
+    }
+
+    public final void onPing() {
+        if (pingState == PingState.SHUTDOWN) {
+            return;
+        }
+
+        if (pingIdleTimeNanos > 0) {
+            firstPingIdleEvent = true;
+            lastPingIdleTime = System.nanoTime();
+        }
+        pingState = PingState.IDLE;
+        cancelFutures();
+    }
+
+    protected abstract ChannelFuture writePing(ChannelHandlerContext ctx);
+
+    protected abstract boolean pingResetsPreviousPing();
+
+    protected abstract boolean hasRequestsInProgress(ChannelHandlerContext ctx);
+
+    @Nullable
+    protected final Future<?> shutdownFuture() {
+        return shutdownFuture;
+    }
+
+    protected final boolean isPendingPingAck() {
+        return pingState == PingState.PENDING_PING_ACK;
+    }
+
+    @VisibleForTesting
+    final PingState state() {
+        return pingState;
+    }
+
+    @VisibleForTesting
+    void onIdleEvent(ChannelHandlerContext ctx, IdleStateEvent evt) {
```

Comment:
How about inlining this? I think connection and ping idle event doesn't have to be in the same place.

Reply:
That sounds good. I will remove this generalization.